### PR TITLE
Enhancements for SAML2 bearer flow

### PR DESCRIPTION
**Pull request by strehle**

### CODE REVIEW

File: server/src/main/java/org/cloudfoundry/identity/uaa/authentication/BackwardsCompatibleTokenEndpointAuthenticationFilter.java [hunk 1]
```diff
@@ -237,9 +238,15 @@ protected Authentication attemptTokenAuthentication(HttpServletRequest request,
...
                     authResult = saml2BearerGrantAuthenticationConverter.convert(request);
+                } catch (AuthenticationException e) {
+                    String errorMessage = (e instanceof Saml2AuthenticationException saml2AuthenticationException) ?
+                            saml2AuthenticationException.getSaml2Error().getDescription() : e.getMessage();
+                    log.debug(errorMessage, e);
+                    throw new InsufficientAuthenticationException(errorMessage);
                 } catch (Exception e) {
-                    log.error("Error setting assertion in SAML filter", e);
-                    throw new InsufficientAuthenticationException("Error setting assertion in SAML filter");
+                    String errorMessage = "Error setting assertion in SAML filter";
```
Comment (strehle) [inline server/src/main/java/org/cloudfoundry/identity/uaa/authentication/BackwardsCompatibleTokenEndpointAuthenticationFilter.java:247]: separated error situations...
In SAML bearer we authenticate a client and thus we can return more details about the problems, e.g. expired SAML assertion ...

File: server/src/main/java/org/cloudfoundry/identity/uaa/provider/saml/ConfiguratorRelyingPartyRegistrationRepository.java [hunk 2]
```diff
@@ -38,19 +36,26 @@ public ConfiguratorRelyingPartyRegistrationRepository(String uaaWideSamlEntityID
     @Override
     public RelyingPartyRegistration findByRegistrationId(String registrationId) {
         IdentityZone currentZone = retrieveZone();
-        List<SamlIdentityProviderDefinition> identityProviderDefinitions = configurator.getIdentityProviderDefinitionsForZone(currentZone);
+        AbstractIdentityProviderDefinition idpDefinition = configurator.getIdentityProviderDefinitionsForOrigin(currentZone, registrationId);
```
Comment (strehle) [inline server/src/main/java/org/cloudfoundry/identity/uaa/provider/saml/ConfiguratorRelyingPartyRegistrationRepository.java:39]: this is the code which resolves IdP by origin or issue

File: server/src/main/java/org/cloudfoundry/identity/uaa/provider/saml/ConfiguratorRelyingPartyRegistrationRepository.java [hunk 3]
```diff
@@ -38,19 +36,26 @@ public ConfiguratorRelyingPartyRegistrationRepository(String uaaWideSamlEntityID
...
+        if (idpDefinition == null) {
+            idpDefinition = configurator.getIdentityProviderDefinitionsForIssuer(currentZone, registrationId);
+        }
+        if (idpDefinition instanceof SamlIdentityProviderDefinition foundSamlIdentityProviderDefinition) {
+            return createRelyingPartyRegistration(foundSamlIdentityProviderDefinition.getIdpEntityAlias(), foundSamlIdentityProviderDefinition, currentZone);
+        }
 
+        List<SamlIdentityProviderDefinition> identityProviderDefinitions = configurator.getIdentityProviderDefinitionsForZone(currentZone);
```
Comment (strehle) [inline server/src/main/java/org/cloudfoundry/identity/uaa/provider/saml/ConfiguratorRelyingPartyRegistrationRepository.java:47]: this loop here is hopefully later not needed

File: server/src/main/java/org/cloudfoundry/identity/uaa/provider/saml/ConfiguratorRelyingPartyRegistrationRepository.java [hunk 4]
```diff
@@ -38,19 +36,26 @@ public ConfiguratorRelyingPartyRegistrationRepository(String uaaWideSamlEntityID
...
         for (SamlIdentityProviderDefinition identityProviderDefinition : identityProviderDefinitions) {
-            if (identityProviderDefinition.getIdpEntityAlias().equals(registrationId)) {
-                return createRelyingPartyRegistration(registrationId, identityProviderDefinition, currentZone);
+            if (registrationId.equals(identityProviderDefinition.getIdpEntityAlias()) || registrationId.equals(identityProviderDefinition.getIdpEntityId())) {
+                return createRelyingPartyRegistration(identityProviderDefinition.getIdpEntityAlias(), identityProviderDefinition, currentZone);
             }
         }
 
-        if (!identityProviderDefinitions.isEmpty()) {
+        // TODO remove hack
```
Comment (strehle) [inline server/src/main/java/org/cloudfoundry/identity/uaa/provider/saml/ConfiguratorRelyingPartyRegistrationRepository.java:54]: and this is the problematic part... without it breaks the SAMLIT - idpInitated SSO

File: server/src/main/java/org/cloudfoundry/identity/uaa/provider/saml/Saml2BearerGrantAuthenticationConverter.java [hunk 5]
```diff
@@ -184,36 +186,26 @@ static Converter<OpenSaml4AuthenticationProvider.AssertionToken, AbstractAuthent
...
-        RelyingPartyRegistration relyingPartyRegistration = relyingPartyRegistrationResolver.resolve(request, null);
 
         String serializedAssertion = request.getParameter("assertion");
-        byte[] decodedAssertion = Saml2Utils.samlDecode(serializedAssertion);
+        byte[] decodedAssertion = Saml2Utils.samlBearerDecode(serializedAssertion);
         String assertionXml = new String(decodedAssertion, StandardCharsets.UTF_8);
 
         Assertion assertion = parseAssertion(assertionXml);
+        RelyingPartyRegistration relyingPartyRegistration = relyingPartyRegistrationResolver.resolve(request, getIssuer(assertion));
+        IdentityProvider<SamlIdentityProviderDefinition> idp = retrieveSamlIdpSamlIdentityProvider(relyingPartyRegistration.getRegistrationId());
```
Comment (strehle) [inline server/src/main/java/org/cloudfoundry/identity/uaa/provider/saml/Saml2BearerGrantAuthenticationConverter.java:196]: this change here simply ensures we have the IDP (and active) before we validate the signature.

File: server/src/main/java/org/cloudfoundry/identity/uaa/provider/saml/ConfiguratorRelyingPartyRegistrationRepository.java (hunk 4 above)
Comment (strehle) [reply server/src/main/java/org/cloudfoundry/identity/uaa/provider/saml/ConfiguratorRelyingPartyRegistrationRepository.java:54]: Not needed with https://github.com/cloudfoundry/uaa/pull/3136

**Review by duanemay — APPROVED** [review]

**Pull request merged**